core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -993,28 +1019,40 @@ public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValu
         if (preValue.bytes() != shardIndexingBufferSize.bytes()) {
             // so we push changes these changes down to IndexWriter:
             engine.onSettingsChanged();
+            logger.debug("updating index_buffer_size from [{}] to [{}]", preValue, shardIndexingBufferSize);
+
+            long iwBytesUsed = engine.indexWriterRAMBytesUsed();
+            if (shardIndexingBufferSize.bytes() < iwBytesUsed) {
+                // our allowed buffer was changed to less than we are currently using; we ask IW to refresh
+                // so it clears its buffers (otherwise it won't clear until the next indexing/delete op)
+                logger.debug("refresh because index buffer decreased to [{}] and IndexWriter is now using [{}] bytes",
+                             shardIndexingBufferSize, iwBytesUsed);
 
-            if (shardIndexingBufferSize == EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER) {
-                // it's inactive: make sure we do a refresh / full IW flush in this case, since the memory
-                // changes only after a "data" change has happened to the writer
-                // the index writer lazily allocates memory and a refresh will clean it all up.
-                logger.debug("updating index_buffer_size from [{}] to (inactive) [{}]", preValue, shardIndexingBufferSize);
+                // TODO: should IW have an API to move segments to disk, but not refresh?
                 try {
                     refresh("update index buffer");
                 } catch (Throwable e) {
-                    logger.warn("failed to refresh after setting shard to inactive", e);
+                    logger.warn("failed to refresh after decreasing index buffer", e);
                 }
-            } else {
-                logger.debug("updating index_buffer_size from [{}] to [{}]", preValue, shardIndexingBufferSize);
             }
         }
 
         engine.getTranslog().updateBuffer(shardTranslogBufferSize);
     }
 
+    /** Record that this shard is now inactive, and decrease the indexing and translog buffers to tiny values. */
     public void markAsInactive() {
-        updateBufferSize(EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER, TranslogConfig.INACTIVE_SHARD_TRANSLOG_BUFFER);
-        indicesLifecycle.onShardInactive(this);
+        if (active.getAndSet(false)) {
+            updateBufferSize(EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER, TranslogConfig.INACTIVE_SHARD_TRANSLOG_BUFFER);
+            logger.debug("shard is now inactive");
+            indicesLifecycle.onShardInactive(this);
+        }
+    }
+
+    /** Returns {@code true} if this shard is active (has seen indexing ops in the last {@link
+     *  IndexingMemoryController#SHARD_INACTIVE_TIME_SETTING} (default 5 minutes), else {@code false}. */
```

Comment:
I wonder if it now makes more sense to have this setting on a shard level (since shard owns the inactivity). We can fall back to the old, node level settings (or just remove it if this is 3.0 change only and deprecate on 2.x)